From 6f0aa92473456b680ddd35421c15f6f41b97c998 Mon Sep 17 00:00:00 2001
From: Jeff Layton <jlayton@redhat.com>
Date: Mon, 21 Apr 2014 12:25:10 -0400
Subject: [fs] nfs: emit a fsnotify_nameremove call in sillyrename codepath

Message-id: <1398083110-3724-6-git-send-email-jlayton@redhat.com>
Patchwork-id: 78983
O-Subject: [RHEL7.1 PATCH 5/5] [BZ#919347] nfs: emit a fsnotify_nameremove call in sillyrename codepath
Bugzilla: 919347
RH-Acked-by: Steve Dickson <SteveD@redhat.com>
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>

If a file is sillyrenamed, then the generic vfs_unlink code will skip
emitting fsnotify events for it.

This patch has the sillyrename code do that instead.

In truth this is a little bit odd since we aren't actually removing the
dentry per-se, but renaming it. Still, this is probably the right thing
to do since it's what userland apps expect to see when an unlink()
occurs or some file is renamed on top of the dentry.

Signed-off-by: Jeff Layton <jlayton@redhat.com>
Tested-by: Anna Schumaker <Anna.Schumaker@netapp.com>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=919347
Upstream commit: f7be728468263fcbaa1e9dcae83fb97a88b4127c

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/unlink.c b/fs/nfs/unlink.c
index 818ded7..de54129 100644
--- a/fs/nfs/unlink.c
+++ b/fs/nfs/unlink.c
@@ -14,6 +14,7 @@
 #include <linux/sched.h>
 #include <linux/wait.h>
 #include <linux/namei.h>
+#include <linux/fsnotify.h>
 
 #include "internal.h"
 #include "nfs4_fs.h"
@@ -465,8 +466,18 @@ nfs_async_rename(struct inode *old_dir, struct inode *new_dir,
 static void
 nfs_complete_sillyrename(struct rpc_task *task, struct nfs_renamedata *data)
 {
- if (task->tk_status != 0)
-  nfs_cancel_async_unlink(data->old_dentry);
+ struct dentry *dentry = data->old_dentry;
+
+ if (task->tk_status != 0) {
+  nfs_cancel_async_unlink(dentry);
+  return;
+ }
+
+ /*
+  * vfs_unlink and the like do not issue this when a file is
+  * sillyrenamed, so do it here.
+  */
+ fsnotify_nameremove(dentry, 0);
 }
 
 #define SILLYNAME_PREFIX ".nfs"
-- 
1.7.1